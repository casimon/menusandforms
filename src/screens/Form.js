import React from 'react'
import { useParams } from 'react-router-dom'
import Button from '@material-ui/core/Button'
import { makeStyles } from '@material-ui/core/styles'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'
import Container from '@material-ui/core/Container'
import Grid from '@material-ui/core/Grid'

import mockData from '../data/mockData.json'

const useStyles = makeStyles((theme) => ({
  root: {
    flexGrow: 1,
    padding: theme.spacing(2),
  },
}))

function Page() {
  let { id } = useParams()
  const classes = useStyles()

  const form = mockData.forms.byId[id]
  const { title, fields, submitUrl, cancelUrl } = form || {}

  const handleSubmit = (event) => {
    event.preventDefault()
    const _form = event.target
    const data = new FormData(_form)
    const body = JSON.stringify(Object.fromEntries(data))
    alert(body)
    fetch(submitUrl, {
      method: 'POST',
      body,
    })
      .then((data) => {
        console.log('Success:', data)
      })
      .catch((error) => {
        console.error('Error:', error)
      })
  }

  return (
    <div className={classes.root}>
      <Container maxWidth='sm'>
        <Grid
          container
          direction='column'
          justify='center'
          alignItems='stretch'
          spacing={3}
        >
          <Grid item xs={12}>
            <Typography variant='h5' component='h1' gutterBottom>
              {title}
            </Typography>
          </Grid>
          {fields.length > 0 && (
            <form onSubmit={handleSubmit}>
              <Grid container spacing={3}>
                {fields.map((fieldId, index) => {
                  const field = mockData.fields.byId[fieldId]
                  const { dataParse, required, label, type } = field || {}
                  return (
                    <Grid item xs={12} md={6} key={index}>
                      <TextField
                        required={required}
                        label={label}
                        type={type}
                        name={fieldId}
                        data-parse={dataParse}
                        fullWidth
                      />
                    </Grid>
                  )
                })}
                <Grid item xs={12}>
                  <Button
                    type='submit'
                    variant='contained'
                    color='primary'
                    fullWidth
                  >
                    Submit
                  </Button>
                </Grid>
                <Grid item xs={12}>
                  <Button
                    href={cancelUrl}
                    variant='contained'
                    color='primary'
                    fullWidth
                  >
                    Cancel
                  </Button>
                </Grid>
              </Grid>
            </form>
          )}
        </Grid>
      </Container>
    </div>
  )
}

export default Page
