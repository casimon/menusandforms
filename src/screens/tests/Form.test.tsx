import React from 'react'

import { render } from '@testing-library/react'
import Form from '../Form'
jest.mock('react-router-dom', () => ({
  useParams: () => ({
    id: 1,
  }),
}))
const init = () => render(<Form />)

describe('My Form component', () => {
  test('should render form 1 ', () => {
    const { container } = init()
    expect(container).toMatchSnapshot()
  })
})
