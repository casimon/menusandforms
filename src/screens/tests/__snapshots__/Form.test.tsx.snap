// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My Form component should render form 1  1`] = `
<div>
  <div
    class="makeStyles-root-1"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthSm"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-direction-xs-column MuiGrid-justify-xs-center"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <h1
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
          >
            You chose well. Fill in your details to win!
          </h1>
        </div>
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
                data-parse="uppercase"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink="false"
                >
                  Name
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    name="userName"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                  data-shrink="false"
                >
                  Email
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    name="email"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <a
                aria-disabled="false"
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
                href="/menu/1"
                tabindex="0"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </a>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
